Plutarch/DataRepr.hs

```diff
@@ -35,60 +40,85 @@ import Plutarch.Builtin (
   pfstBuiltin,
   psndBuiltin,
  )
+import Plutarch.Field.HList (type Drop, type IndexList)
 import Plutarch.Integer (PInteger)
 import Plutarch.Lift (PConstant, PConstantRepr, PConstanted, PLift, pconstantFromRepr, pconstantToRepr)
-import Plutarch.List (punsafeIndex)
+import Plutarch.List (pdrop, punsafeIndex, punsafeIndex')
 import Plutarch.Prelude
 import qualified Plutus.V1.Ledger.Api as Ledger
 import qualified PlutusCore as PLC
 
-data PDataList (as :: [PType]) (s :: S)
+data PDataRecord (as :: [PLabeled]) (s :: S)
 
-pdhead :: Term s (PDataList (a : as) :--> PAsData a)
+data PLabeled = Symbol := PType
+
+pdhead :: Term s (PDataRecord ((l ':= a) : as) :--> PAsData a)
 pdhead = phoistAcyclic $ pforce $ punsafeBuiltin PLC.HeadList
 
-pdtail :: Term s (PDataList (a : as) :--> PDataList as)
+pdtail :: Term s (PDataRecord (a ': as) :--> PDataRecord as)
 pdtail = phoistAcyclic $ pforce $ punsafeBuiltin PLC.TailList
 
-type PDataRepr :: [[PType]] -> PType
-data PDataRepr (defs :: [[PType]]) (s :: S)
+type family PTypes (as :: [PLabeled]) :: [PType] where
+  PTypes '[] = '[]
+  PTypes ((l ':= t) ': as) = t ': (PTypes as)
+
+type family PNames (as :: [PLabeled]) :: [Symbol] where
+  PNames '[] = '[]
+  PNames ((l ':= t) ': as) = l ': (PNames as)
+
+type PDataRepr :: [[PLabeled]] -> PType
+data PDataRepr (defs :: [[PLabeled]]) (s :: S)
 
 pasData :: Term s (PDataRepr _) -> Term s PData
 pasData = punsafeCoerce
 
-type family IndexList (n :: Nat) (l :: [k]) :: k where
-  IndexList 0 (x ': _) = x
-  IndexList n (x : xs) = IndexList (n - 1) xs
-
-punDataRepr :: Term s (PDataRepr '[def] :--> PDataList def)
+punDataRepr :: Term s (PDataRepr '[def] :--> PDataRecord def)
 punDataRepr = phoistAcyclic $
   plam $ \t ->
     plet (pasConstr #$ pasData t) $ \d ->
-      (punsafeCoerce $ psndBuiltin # d :: Term _ (PDataList def))
+      (punsafeCoerce $ psndBuiltin # d :: Term _ (PDataRecord def))
 
-pindexDataRepr :: (KnownNat n) => Proxy n -> Term s (PDataRepr (def : defs) :--> PDataList (IndexList n (def : defs)))
+pindexDataRepr :: (KnownNat n) => Proxy n -> Term s (PDataRepr (def : defs) :--> PDataRecord (IndexList n (def : defs)))
```

Comment:
I like `PDataSum`

Reply:
@t1lde Wouldn't it be `PDataProduct`, because you are using this in products - such as `PTxInfo` fields, and the arguments of a sum constructor like `PMinting`, rather than sum types per se (for which we use Haskell ADT syntax).

It is the "P" in "SOP" (sum of products) of `generics-sop`.

Reply:
`PDataSum` is a replacement for `PDataRepr`.